### ISSUE TITLE
APPSODY_PREP is not run before any of the ON_CHANGE commands.

### ISSUE THREAD

**Issue by henrynash** [appsody/appsody#352]
Counter to what our documentation says, APPSODY_PREP is never executed when a command is executed by the watcher. It is only run ahead of the initial RUN/TEST/DEBUG on launch of the stack.

This means you cannot, for example, inject dependencies if they are added while in Rapid Local Development mode without killing your appsody run and restarting.

**Comment by kewegner** [comment]
re: Documentation.  I think what it means (maybe poorly stated) is that APPSODY_PREP is run prior to running any RUN/TEST/Debug and ON_CHANGE actions.
note:
Unlike those commands, APPSODY_PREP will only be run once and never retried. 

As far as running it before every on_change command, we'd have to discuss that with the wider team (cli and stacks) as the behavior you see was indeed how it was implemented.  We can discuss changing it.

**Comment by henrynash** [comment]
OK, got it - that's why I founds the docs confusing, since I don't actually think it is ever run ahead of ON_CHANGE (since isn't ON_CHANGE only ever be run after the relevant initial RUN/TEST/DEBUG is executed on launch)?

Agreed, that we would need to think carefully about changing this, as it doesn't do that now.

**Comment by henrynash** [comment]
I think probably the best action right now is to clarify the docs (I'll happily put in a separate PR for that to the website).

In retrospect, I suspect we actually probably DON'T want, by default, to run PREP before every ON_CHANGE...I worry about slowing down the fast spin up of the app (and changing dependencies will be a rare event compared to regular file changes). As you say, you can encode this in the ON_CHANGE var yourself - e.g. I just tried this out in the python-flask stack:

ENV APPSODY_RUN_ON_CHANGE="( $APPSODY_PREP ; python -m flask run --host=0.0.0.0 --port=8080 )"

**Comment by kylegc** [comment]
I agree, lets get the docs updated. The PREP command is just that, preparation steps before you get into the run/on_change cycle. It shouldn't be run on every on_change.

@henrynash thanks for volunteering to fix the docs, I'll assign this issue to you.
